File: java/src/main/java/io/binac/leetcode/UniqueBinarySearchTreesII.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * <p>Given an integer <code>n</code>, return <em>all the structurally unique <strong>BST'</strong>s (binary search trees), which has exactly </em><code>n</code><em> nodes of unique values from</em> <code>1</code> <em>to</em> <code>n</code>. Return the answer in <strong>any order</strong>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2021/01/18/uniquebstn3.jpg" style="width: 600px; height: 148px;">
 * <pre><strong>Input:</strong> n = 3
 * <strong>Output:</strong> [[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> n = 1
 * <strong>Output:</strong> [[1]]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= n &lt;= 8</code></li>
 * </ul>
 */
public class UniqueBinarySearchTreesII {
    public static class Solution1 {
        private TreeNode cloneTree(TreeNode root) {
            if (root == null)
                return null;
            return new TreeNode(root.val, cloneTree(root.left), cloneTree(root.right));
        }

        private List<TreeNode> generateTrees(int start, int end) {
            if (start == end)
                return Collections.singletonList(null);
            List<TreeNode> result = new ArrayList<>();
            for (int i = start; i < end; ++i) {
                List<TreeNode> lefts = generateTrees(start, i);
                List<TreeNode> rights = generateTrees(i + 1, end);
                for (TreeNode left : lefts) {
                    for (TreeNode right : rights) {
                        result.add(new TreeNode(i, cloneTree(left), cloneTree(right)));
                    }
                }
            }
            return result;
        }

        public List<TreeNode> generateTrees(int n) {
            return generateTrees(1, n + 1);
        }
    }
}

File: python3/leetcodepy/unique_binary_search_trees_ii.py
"""
Given an integer n, return all the structurally unique BST's (binary search trees), which has exactly n nodes of unique values from 1 to n. Return the answer in any order.



Example 1:
https://assets.leetcode.com/uploads/2021/01/18/uniquebstn3.jpg

Input: n = 3
Output: [[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]

Example 2:

Input: n = 1
Output: [[1]]


Constraints:

1 <= n <= 8
"""
from typing import List, Optional

from .utils import TreeNode


class Solution1:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.generate_trees(1, n + 1)

    def copy_tree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None
        return TreeNode(root.val, self.copy_tree(root.left), self.copy_tree(root.right))

    def generate_trees(self, start: int, end: int) -> List[Optional[TreeNode]]:
        if start == end:
            return [None]
        result = []
        for i in range(start, end):
            left_trees = self.generate_trees(start, i)
            right_trees = self.generate_trees(i + 1, end)
            for left_tree in left_trees:
                for right_tree in right_trees:
                    result.append(TreeNode(i, self.copy_tree(left_tree), self.copy_tree(right_tree)))
        return result
